File: Leetcode/119.杨辉三角-ii.cpp
/*
 * @lc app=leetcode.cn id=119 lang=cpp
 *
 * [119] 杨辉三角 II
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
/*2021-07-25-20-22*/
// ? 思路解析：
/* 
 * 解法一： 打表
 * 解法二： 二维数组存储整个三角，输出
 * 
 * 
 ! --难点：
 ! --
 ! --
 */
// @lc code=start
class Solution {
public:
    vector<int> getRow(int rowIndex) {
        
    }
};
// @lc code=end

class classicYangHui {
   public:
    vector<vector<int>> generate(int numRows) {
        //构造一个正方形的二维数组
        vector<vector<int>> result(numRows, vector<int>(numRows, 0));
        //将相应的数字先填到靠左侧的三角形中
        for (int i = 0; i < numRows; i++) {
            result.at(i).at(0) = 1;  //初始化三角的左侧的边界值
            result.at(i).at(i) = 1;  //初始化三角的右侧的边界值
        }
        //内容的计算
        for (int i = 1; i < numRows; i++) {
            for (int j = 1; j < numRows; j++) {
                result[i][j] = result[i - 1][j - 1] + result[i - 1][j];
            }
        }
        //最后将等于 0 的单位(空白单位)弹出数组的空间
        for (int i = 0; i < numRows; i++) {
            for (int j = numRows - i - 1; j > 0; j--) {
                result.at(i).pop_back();
            }
        }
        //返回结果
        return result;
    }
};

File: Leetcode/119.杨辉三角-ii.go
/*
 * @lc app=leetcode.cn id=119 lang=golang
 *
 * [119] 杨辉三角 II
 */

package leetcode

/* --- 2021-10-31-14-33 --- */

// https://github.com/fengwei2002/Algorithm
// solution link:
// https://leetcode-cn.com/problems/pascals-triangle-ii/solution/lc119-fengwei2002-by-kycu-nfcf/

// @lc code=start
func getRow(rowIndex int) []int {
	f := make([][]int, 2)
	for i := 0; i < 2; i++ {
		f[i] = make([]int, rowIndex+1)
	}
	// 首先不用滚动数组解决这个问题，然后对数组的行 index 进行修改
	// 第 0 行存储 0 2 4 6 8 10 行
	// 第 1 行存储 1 3 5 7 9 行
	// 第 n 行应该存储到 n % 2 行，也就是 & 1 的地方

	for i := 0; i <= rowIndex; i++ {
		f[i&1][0], f[i&1][i] = 1, 1
		for j := 1; j < i; j++ {
			f[i&1][j] = f[(i-1)&1][j-1] + f[(i-1)&1][j]
		}
	}
	return f[rowIndex&1]
}

// @lc code=end
